### PR TITLE
Finish migrating from email to sub for auth

### LINKED ISSUE #2722: Finish Migrating From Email To Sub For Authentication
When our schema and user editor was being developed we were using Laravel auth and made three assumptions:
- Users would be authenticated by their email address
- Every user would have an email address
- Email address would never change

Now that we're using SiC these assumptions are not longer valid.  Instead:
- Users are authenticated by `sub` provided by the oauth provider
- Not every user can log in

Now we need to finish this migration:
- [ ] Email address can be blank and edited
- [ ] Sub can be blank and edited

Check:
- [ ] User Model
- [ ] Schema
- [ ] Admin forms
## Acceptance Criteria
- [ ] A user's `email` and `sub` values should be editable and nullable
  - [ ] A user can be created with no `email` or `sub` in the admin interface
  - [ ] An admin should be able to fill these values
  - [ ] An admin should be able to edit these values
  - [ ] An admin should be able to clear these values
  - [ ] An autocreated user should have the `sub` but no `email`
  - [ ] An autocreated user should be able to log in but have no roles and not be able access protected pages
  - [ ] An admin should be able to fill in the fields of an autocreated user to promote them to admin
  - [ ] Login should work with SiC
  - [ ] Login should work with Laravel Auth or Mock Auth (whatever is going by then :smile:)

## Estimated Tasks
Note, some of these are probably already done. Part of this task is double checking that.
- [ ] Update database with new migration
- [ ] Update eloquent models, factories, seeders
- [ ] Update schema queries, inputs, mutations
- [ ] Update Admin User Create and Edit forms

### PR DESCRIPTION
**Pull request by petertgiles**
This branch finishes migrating from `email` to `sub` for authentication.
- Updates to schema
- Updates to admin site
- Updates to profile site

Use cases for testing
- [ ] A user can be created with no email or sub in the admin interface
- [ ] An admin should be able to fill these values
- [ ] An admin should be able to edit these values
- [ ] An admin should be able to clear these values
- [ ] An autocreated user should have the sub but no email
- [ ] An autocreated user should be able to log in but have no roles and not be able access protected pages
- [ ] An admin should be able to fill in the fields of an autocreated user to promote them to admin
- [ ] Login should work with SiC
- [ ] Login should work with Laravel Auth
- [ ] Regular (non-admin) user can update their email in *profiles*

Closes #2722

### CODE REVIEW

**Comment by petertgiles** [comment]
Chromatic failures are addressed in #2876

File: api/app/Providers/AuthServiceProvider.php
Comment (tristan-orourke) [inline api/app/Providers/AuthServiceProvider.php:112]: I'm not sure if this is an improvement? Since this setting is only for convenience during local development, I think using an email is more intuitive. I can set `AUTH_DEFAULT_USER=admin@test.com` without having to look up the sub value.

File: api/graphql/schema.graphql
Comment (patcon) [reply api/graphql/schema.graphql:597]: +1 to keeping the uniqueness of emails. Even though we don't technically need it, feels really against the grain to allow multiple emails for accounts, without us kinda "clearing" that as sensible.

The value of releasing a constrain doesn't feel worth losing the sanity-check :) Seems like it could get messy

**Comment by patcon** [comment]
Just a gut-check, but will our seeded admit still be able to have an email as a `sub`? 

(Just in case, would love to merge #2813 before this, as that PR is prone to merge conflicts due to removing all of `auth`, and it's holding up the cypress PR)

File: api/app/Providers/AuthServiceProvider.php
Comment (petertgiles) [reply api/app/Providers/AuthServiceProvider.php:112]: I changed this because it is the last remaining spot that `email` is used in an auth context so it is very out of place.  The seeder also seeds "admin@test.com" as the `sub` for this user so you don't have to look up a new value - it's the same. :smile:

**Comment by petertgiles** [comment]
> would love to merge #2813 before this

No worries.  We've got a week left in the sprint and this isn't blocking anything.

**Comment by petertgiles** [comment]
> will our seeded admit still be able to have an email as a `sub`?

Yeah, there are no restrictions on sub other than it must be unique.  We can seed `test` or `000` or whatever.  I left our "dummy admin" with the sub `admin@test.com` since that's what people are used to.

File: api/graphql/schema.graphql
Comment (petertgiles) [reply api/graphql/schema.graphql:597]: I removed it because it seemed like this special auth-specific handling.  We don't have the constraint on `telephone`, for example.  

We actually still have the database constraint in place for email so it is enforcing uniqueness there.  Is this directive redundant?  Should it be moved to sub?  Should we have the directive on both fields so they match the database?  Should we remove the database constraint on email altogether?  Will we have edge cases where multiple users want to use a shared mailbox or distro list?  If they end up with multiple accounts that need to be merged matching on email will be a nice trick.

I agree that this is inconsistent as it is but I'm not sure where we want to fall on consistency.

File: frontend/admin/src/js/components/user/CreateUser.tsx
Comment (petertgiles) [reply frontend/admin/src/js/components/user/CreateUser.tsx]: Good call!  I've moved it to util.

File: api/graphql/schema.graphql
Comment (tristan-orourke) [reply api/graphql/schema.graphql:597]: 1. I believe our schema rules should match the database. I think it makes the constraints (and the errors if they're violated) more visible.
2. I'm leaning towards keeping the email uniqueness constraint. Eventually we may probably want to confirm emails, and if emails are unique that would help avoid spam accounts. Merging accounts should be a rare and exceptional event, and Admins could modify the old account if needing to free up the email again.

Comment (petertgiles) [reply api/graphql/schema.graphql:597]: Yeah, I like that.  I've added the schema validation back in for email and added one for sub as well.

Comment (patcon) [reply api/graphql/schema.graphql:597]: > I'm leaning towards keeping the email uniqueness constraint. Eventually we may probably want to confirm emails

Similar vibes. It feels possible that once running, we might learn about some pragmatic downsides to treating email as unverified contact details (it seems like a very ubiquitous pattern, even when proliferation of oauth providers mean it's not required). Cautiousness would seem to avoid heavier corrective interactions with actual users, if that's needed (which is the sort of thing that might paralyze a decision 🙂 )

**Comment by patcon** [comment]
Screen Shot 2022-05-26 at 12 04 56 PM

I understand that this is technically a valid email, but do we want this to work without dots? (seems a bit more relevant since we don't validate). Maybe beyond scope, so happy to reticket or bury :)

**Comment by patcon** [comment]
Is making `sub` editable something we've been asked for, or something we added in anticipation of it helping for future process?

I'm trying to foresee where things could go bad with allowing `sub` to be changed so easily? E.g., someone's cat walks on the keyboard an removes a letter from a uuid. Without an email, and without database change history, i don't think there's an easy way to re-attach, right? Would require going to a database dump or getting on the phone with GCKey folks, right? Maybe I'm lacking imagination, but seems more risky than necessary to put it right there with other fields that are really safe to change

Also, if we do keep it "Subject" seems like it's a term leaking through the auth protocol abstraction. Maybe something more like "External ID"

**Comment by patcon** [comment]
Also, maybe this is over-thinking, but it also breaks a piece of any ad-hoc audit trail we might have. If any admin could swap their ID to someone else's, then it breaks one of the more authoritative assurances we have (that a GCKey account using a gov email is logging in and making some change)

**Review by patcon — APPROVED** [review]
LGTM

**Comment by petertgiles** [comment]
Thanks for the review!  Good questions about sub - I've started #2887 to capture your ideas.  Feel free to add more over there.  Regarding emails - we do have validation in the input field and in the graphql layer.  If you feel like you want to tighten it up further you could start an issue for that, too.

**Review by tristan-orourke — APPROVED** [review]
@patcon those are some valid questions about sub, @petertgiles thanks for making a new issue. I think this one's good for now!

**Pull request merged**
